lib/HLSL/DxilModule.cpp

```diff
@@ -367,7 +389,65 @@ void DxilModule::CollectShaderFlags(ShaderFlags &Flags) {
           case DXIL::OpCode::Msad:
             hasMSAD = true;
             break;
-          case DXIL::OpCode::BufferLoad:
+          case DXIL::OpCode::BufferLoad: {
+              Value *resHandle = CI->getArgOperand(DXIL::OperandIndex::kBufferStoreHandleOpIdx);
+              CallInst *handleCall = cast<CallInst>(resHandle);
+
+              if (ConstantInt *resClassArg =
+                  dyn_cast<ConstantInt>(handleCall->getArgOperand(
+                      DXIL::OperandIndex::kCreateHandleResClassOpIdx))) {
+                  DXIL::ResourceClass resClass = static_cast<DXIL::ResourceClass>(
+                      resClassArg->getLimitedValue());
+                  if (resClass == DXIL::ResourceClass::UAV) {
+                    if (ConstantInt *resRangeID =
+                            dyn_cast<ConstantInt>(handleCall->getArgOperand(
+                                DXIL::OperandIndex::kCreateHandleResIDOpIdx))) {
+                      DxilResource &res = GetUAV(resRangeID->getLimitedValue());
+                      if (res.GetKind() == DXIL::ResourceKind::TypedBuffer) {
+                        Value *GV = res.GetGlobalSymbol();
+                        llvm::Type *Ty = GV->getType()->getPointerElementType();
+                        // Check if uav load is multi component load.
+                        if (!IsResourceSingleType(Ty)) {
+                          hasMulticomponentUAVLoads = true;
+                        }
+                      }
+                    }
+                  }
+              }
+              else if (PHINode *resClassPhi = dyn_cast<
+                  PHINode>(handleCall->getArgOperand(
+                      DXIL::OperandIndex::kCreateHandleResClassOpIdx))) {
+                  unsigned numOperands = resClassPhi->getNumOperands();
+                  for (unsigned i = 0; i < numOperands; i++) {
+                      if (ConstantInt *resClassArg = dyn_cast<ConstantInt>(
+                          resClassPhi->getIncomingValue(i))) {
+                          DXIL::ResourceClass resClass =
+                              static_cast<DXIL::ResourceClass>(
+                                  resClassArg->getLimitedValue());
+                          if (resClass == DXIL::ResourceClass::UAV) {
+                            if (ConstantInt *resRangeID = dyn_cast<ConstantInt>(
+                                    handleCall->getArgOperand(
+                                        DXIL::OperandIndex::
+                                            kCreateHandleResIDOpIdx))) {
```

Comment:
If resource class is going through a phi, couldn't resource ID also be going through a phi?  There are no phi test cases that I can see, though there may be some other tests that hit this code path.